src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,51 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return BATTERY_AMP_HOUR_CAPACITY-ampHours ;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
+    if(previousCurrent<0)
+    {
     return false;
+    }
+    else
+    {
+       return true;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+
+    double timeRemaining = (BATTERY_AMP_HOUR_CAPACITY-ampHours)/previousCurrent ;
+
+    timeRemaining = qAbs(timeRemaining)/MILISECONDS_TO_HOURS;
+
+    return QTime(0, 0, 0, 0).addMSecs(timeRemaining);
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    double newCurrent = batteryData.current;
+    if(previousCurrent == 0)
+    {
+        aveCurrent = newCurrent;
+    }
+    else
+    {
+        aveCurrent = (previousCurrent+newCurrent)/2;
+    }
+
+    newTime = batteryData.time;
+    deltaTime = previousTime.msecsTo(newTime);
+
+    deltaTime = deltaTime*MILISECONDS_TO_HOURS;
+
+    double dAmpHours = aveCurrent*deltaTime;
```

Comment:
"dAmpHours" is pretty ambiguous, change it to something more understandable like "ampHoursBetweenPrevious_PresentTime". Long variable names are ok!